modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C

```diff
@@ -47,16 +55,37 @@ ComputeFiniteStrainElasticStress::computeQpStress()
   // Calculate the stress in the intermediate configuration
   RankTwoTensor intermediate_stress;
 
-  intermediate_stress =
-      _elasticity_tensor[_qp] * (_elastic_strain_old[_qp] + _strain_increment[_qp]);
+  if (hasGuaranteedMaterialProperty(_elasticity_tensor_name, Guarantee::ISOTROPIC))
+  {
+    intermediate_stress =
+        _elasticity_tensor[_qp] * (_elastic_strain_old[_qp] + _strain_increment[_qp]);
+
+    // Compute dstress_dstrain
+    _Jacobian_mult[_qp] = _elasticity_tensor[_qp]; // This is NOT the exact jacobian
+  }
+  else
+  {
+    // Rotate elasticity tensor to the intermediate configuration
+    // That is, elasticity tensor is defined in the previous time step
+    // This is consistent with the definition of strain increment
+    // The stress is projected onto the current configuration a few lines below
+    RankFourTensor elasticity_tensor_rotated = _elasticity_tensor[_qp];
+    elasticity_tensor_rotated.rotate(_rotation_total_old[_qp]);
+
+    intermediate_stress =
+        elasticity_tensor_rotated * (_elastic_strain_old[_qp] + _strain_increment[_qp]);
+
+    // Update current total rotation matrix to be used in next step
+    _rotation_total[_qp] = _rotation_increment[_qp] * _rotation_total_old[_qp];
```

Comment:
Surprisingly this is the correct order of multiplication here. I would have thought 
```
    _rotation_total[_qp] *= _rotation_increment[_qp];
```

which swaps the operands on the `*` operator was the way to go. I added a unit test:

```
TEST(RankFourTensor, rotation)
{
  std::vector<Real> input(81);
  for (size_t i = 0; i < input.size(); ++i)
    input[i] = i + 1.1;

  RankFourTensor a(input, RankFourTensor::general);

  RotationTensor xrot(RotationTensor::XAXIS, 90);
  RotationTensor yrot(RotationTensor::YAXIS, 90);

  auto xyrot = xrot * yrot;
  auto yxrot = yrot * xrot;

  auto axy1 = a;
  axy1.rotate(xrot);
  axy1.rotate(yrot);

  auto axy2 = a;
  axy2.rotate(yxrot);

  EXPECT_NEAR(0, (axy1 - axy2).L2norm(), 1E-8);
}
```

How unsatisfactory.

Reply:
This is a great case for unit testing.

For rotations with respect to a global frame, the order of rotations is "inverse": `R_{n}R_{n-1}...R_{1}`, where `n` is the last rigid body rotation. Here, I think we need to get `_rotation_total_old` involved in the computation to keep proper track of what rotation increment corresponds to a converged step.